### PR TITLE
feat: LEAP-1424: Open preview window for images in Grid view

### PR DESCRIPTION
**Pull request by hlomzik**
Open a simple modal with image from the data of clicked task. Allow to navigate between tasks visible in Grid view by Prev/Next buttons. Only one image is displayed right now.

Simple common hotkeys are available:
- `<Arrows left/right>` to switch between tasks
- `<Space>` to select/unselect task
- `<Escape>` to close the modal

Screenshot 2024-12-07 at 19 02 19

### PR fulfills these requirements
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [ ] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

### Change has impacts in these area(s)
- [x] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [x] Frontend

### Feature flags used to cover this change
`fflag_feat_front_leap_1424_grid_preview_short`

### Describe the reason for change
It's hard to quickly preview task and select it depending on a content in current state — if you open task it's another view and it takes time to load it. Here is the new view for quick preview.

### Does this PR introduce a breaking change?
- [x] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [ ] No
- [ ] Not sure (briefly explain the situation below)

Click on task in Grid view now has a different action, but task selection is still possible if you click on task id with checkbox.

### What level of testing was included in the change?
- [ ] e2e
- [ ] integration
- [ ] unit

### CODE REVIEW

File: web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+const TaskModal = observer(({ view, tasks, imageField, currentTaskId, setCurrentTaskId }: TaskModalProps) => {
+  const index = tasks.findIndex(task => task.id === currentTaskId);
+  const task = tasks[index];
+  const src = imageField ? (task?.data?.[imageField] || "") : "";
+
+  const goToNext = () => {
+    if (index < tasks.length - 1) {
+      setCurrentTaskId(tasks[index + 1].id);
+    }
+  };
```
Comment (bmartel) [inline web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx]:
```suggestion
  const goToNext = useCallback(() => {
    if (index < tasks.length - 1) {
      setCurrentTaskId(tasks[index + 1].id);
    }
  }, [tasks]);
```

File: web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+  const goToPrev = () => {
+    if (index > 0) {
+      setCurrentTaskId(tasks[index - 1].id);
+    }
+  };
```
Comment (bmartel) [inline web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx]:
```suggestion
  const goToPrev = useCallback(() => {
    if (index > 0) {
      setCurrentTaskId(tasks[index - 1].id);
    }
  }, [tasks]);
```

File: web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+  const onSelect = () => view.toggleSelected(task.id);
```
Comment (bmartel) [inline web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx]:
```suggestion
  const onSelect = useCallback(() => view.toggleSelected(task?.id), [view.toggleSelected, task?.id]);
```

File: web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+  const onClose = () => setCurrentTaskId(null);
```
Comment (bmartel) [inline web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx]:
```suggestion
  const onClose = useCallback(() => setCurrentTaskId(null), []);
```

File: web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        // pass this event through for other keys
+        return;
+      }
+
+      event.stopPropagation();
+    };
+
+    document.addEventListener("keydown", onKeyDown);
+    return () => document.removeEventListener("keydown", onKeyDown);
+  }, [goToNext, goToPrev]);
```
Comment (bmartel) [inline web/libs/datamanager/src/components/MainView/GridView/GridPreview.tsx]:
```suggestion
  }, [goToNext, goToPrev, onSelect, onClose]);
```

File: web/libs/datamanager/src/components/MainView/GridView/GridView.jsx [hunk 6]
```diff
@@ -56,11 +57,22 @@ const GridDataGroup = observer(({ type, value, field, row }) => {
 });
 
 const GridCell = observer(({ view, selected, row, fields, onClick, ...props }) => {
+  const { setCurrentTaskId, imageField } = useContext(GridViewContext);
+
+  const handleBodyClick = (e) => {
```
Comment (bmartel) [inline web/libs/datamanager/src/components/MainView/GridView/GridView.jsx]: I have seen more times that you have chosen not to wrap an inline function with useCallback than not, can you maybe explain the thinking? Generally speaking it is safer to default to always wrapping with useCallback to ensure there are no stale closures and that it is a proper dependency of any downstream components which are memoized, or as dependencies of other useCallback, useEffect or useMemo.

File: web/libs/datamanager/src/components/MainView/GridView/GridView.jsx [hunk 7]
```diff
@@ -56,11 +57,22 @@ const GridDataGroup = observer(({ type, value, field, row }) => {
...
+    if (!isFF(FF_GRID_PREVIEW) || !imageField) return;
+    // @todo skip this interaction if there are no images in the task
+    e.stopPropagation();
+    setCurrentTaskId(row.id);
+  };
```
Comment (bmartel) [inline web/libs/datamanager/src/components/MainView/GridView/GridView.jsx]:
```suggestion
  const handleBodyClick = useCallback((e) => {
    if (!isFF(FF_GRID_PREVIEW) || !imageField) return;
    // @todo skip this interaction if there are no images in the task
    e.stopPropagation();
    setCurrentTaskId(row.id);
  }, [imageField, row.id]);
```

File: web/libs/datamanager/src/components/MainView/GridView/GridView.jsx (hunk 6 above)
Comment (hlomzik) [reply web/libs/datamanager/src/components/MainView/GridView/GridView.jsx]: the modal is just rendered only once for a task, so I didn't care, but you are right for sure, will update them in a sec

**Review by bmartel — APPROVED** [review]

**Comment by MihajloHoma** [comment]
/git merge 
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/12277603253)
> Successfully merged:  create mode 100644 web/libs/frontend-test/src/helpers/utils/fixLSParams.ts

**Comment by hlomzik** [comment]
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/12282247940)
> Successfully merged: Already up to date.

**Comment by MihajloHoma** [comment]
/git merge


> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/12311351692)
> Successfully merged:  3 files changed, 83 insertions(+), 2 deletions(-)

**Comment by hlomzik** [comment]
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/12355096988)
> Successfully merged:  12 files changed, 699 insertions(+), 40 deletions(-)

**Comment by hlomzik** [comment]
merged already by https://github.com/HumanSignal/label-studio/pull/6781

**Pull request closed without merging**